### DISCUSSION TITLE
Mesa version mismatch causing sway coredump?

### DISCUSSION THREAD

**Discussion by wkral** [Jovian-Experiments/Jovian-NixOS#410] [Q&A] [answered]
I think that because of the custom version of `mesa` in Jovian, when using sway from `nixos-unstable` it is compiled with a different mesa as an input which doesn't match the version that is running on my deck, so when switching to desktop, sway immediately coredumps when trying to launch. It used to work fine running sway and gamescope but hasn't for a long while. I was using an old build for quite a while which was working, I suspect it stopped working around the time mesa 23.x was released but I can't be sure.

I suspect this since searching for the error sway is spitting out: `amdgpu: amdgpu_cs_ctx_create2 failed. (-13)` I found this issue: https://github.com/NixOS/nixpkgs/issues/229108#issuecomment-1529012831 which following that discussion led to: https://github.com/NixOS/nixpkgs/issues/228586#issuecomment-1527030148.

The resolution is not so clear cut however. I tried adding the jovian overlay to nixpkgs just for my steamdeck but it doesn't seem to have changed anything. I thought that might trigger a build of wlroots with the custom mesa version and subsequently sway but it's unclear if they were rebuilt, I didn't notice them appear in the packages it was building afterwards, and thought they would take some time to compile.

Based on what I've found, do you think think the version mismatch is the cause of this crash? If so, how can I make sure the versions match? My config can be found starting here for the deck: https://github.com/wkral/home-nix/blob/de1436c6c879784df75a0a44aa8758f76b34c96a/hosts/default.nix#L33C3-L40C5 and more specifically: https://github.com/wkral/home-nix/blob/de1436c6c879784df75a0a44aa8758f76b34c96a/hosts/deck/configuration.nix

If it's not this any pointers as to what I'm doing wrong would be appreciated. I'm assuming someone has gamescope and sway running at the same time and I must be doing something wrong.

Thanks.

**Comment by K900** [comment]
The Jovian version of Mesa should generally match the nixpkgs one. Are you sure you're not pulling individual packages from unstable or something?

**Reply by wkral** [reply]
I don't think I'm doing anything to override anything coming from Jovian, but sway an many of the other packages are generally coming from unstable as they're not defined in Jovian. I see in the Jovian source that the mesa `version = "24.1.0";` and unstable currently shows `version = "24.2.2";` I don't know how sensitive it is if even minor versions could cause this. Or if this is even the cause. I share a lot of sway config across multiple machines so I'm hoping not to have to do a custom version for the deck.

**Reply by K900** [reply]
You're looking at the wrong Mesa. The right Mesa is in mesa-jupiter, not mesa-radv-jupiter.

**Reply by wkral** [reply]
Sorry, I don't see `mesa-jupiter` file or that name used in the source. I see `mesa-radv-jupiter` and `mesa-radeonsi-jupiter` packages. I'm looking at the development branch, is it possibly somewhere else? The only other mesa thing I can find is the module which references those two. 

I did a `vulkaninfo --summary` earlier which said that radv driver was being used, so that's why I was looking there.

**Reply by K900** [reply]
You want `mesa-radeonsi-jupiter`.

**Reply by wkral** [reply]
Ok, so I see that `mesa-radeonsi-jupiter` just grabs a patch but doesn't change the version. So are you saying that you don't think the sway crash is caused by a version mismatch?

If so is there anything else that could be causing it? I use sway on two other machines which also have AMD GPUs, and they're all set up very similarly, Jovian being the main difference with the steam deck.

**Reply by K900** [reply]
Yes, it's very unlikely to be a version mismatch. In fact, that error is likely a red herring entirely. Do you have a core dump or something?

**Reply by wkral** [reply]
I have several, the main trace where it's failing is as follows:

```
                #0  0x000000000043a470 queue_output_config (sway + 0x3a470)
                #1  0x000000000043afcd apply_output_config (sway + 0x3afcd)
                #2  0x0000000000423498 handle_new_output (sway + 0x23498)
                #3  0x00007fc881a68c4c wl_signal_emit_mutable (libwayland-server.so.0 + 0xac4c)
                #4  0x00007fc881a68c4c wl_signal_emit_mutable (libwayland-server.so.0 + 0xac4c)
                #5  0x00007fc881991a73 scan_drm_connectors (libwlroots.so.12 + 0x5ca73)
                #6  0x00007fc88198ddc3 backend_start (libwlroots.so.12 + 0x58dc3)
                #7  0x00007fc88199cd45 multi_backend_start (libwlroots.so.12 + 0x67d45)
                #8  0x000000000041ea08 server_start (sway + 0x1ea08)
                #9  0x0000000000411fea main (sway + 0x11fea)
                #10 0x00007fc8816d614e __libc_start_call_main (libc.so.6 + 0x2a14e)
                #11 0x00007fc8816d6209 __libc_start_main@@GLIBC_2.34 (libc.so.6 + 0x2a209)
                #12 0x0000000000412205 _start (sway + 0x12205)
```
In the journal it's always preceeded by the `amdgpu: amdgpu_cs_ctx_create2 failed. (-13)` error. There is a segfault and the kernel spits out a hexadecimal code, then the coredump. It's been the same in all of the ones I've looked at.

I have `kanshi` setup to change the output config as a systemd user service. I don't see any mention of it in the journal. Since sway doesn't finish starting, it doesn't trigger kanshi to launch.

I could upload the coredump somewhere if you want the whole file. Do you have a recommended place for that?

Btw, I appreciate your help on this. Thanks.

**Reply by K900** [reply, marked as answer]
Yeah, that's not a Mesa mismatch. $5 says it's https://github.com/swaywm/sway/issues/8348

**Reply by wkral** [reply]
Thank you very much, I definitely got too focused on the red hearing. Setting the background explicitly allows it to launch. I just wonder why this hasn't shown up on my other machines since they have the same sway config/version?

**Reply by K900** [reply]
No idea, you'd have to ask the Sway people.
